### PR TITLE
[FLINK-14290] Decouple plan translation from job execution/ClusterClient

### PR DESCRIPTION
**Pull request by aljoscha**
## What is the purpose of the change

This introduces `Pipeline` as a new common base class of `Plan` (DataSet API) and `StreamGraph` (DataStream API). Then we add `PlanTranslationUtil` which can translate a `Pipeline` to a `JobGraph` for use in the `flink-clients` package. As a last step we change the different environments to all use the `PlanExecutor` abstraction which internally uses the `PipelineTranslationUtil`.

## Brief change log

The commits messages also clearly outline the changes and describe the rationale behind the changes, so looking at them one after the other would be easiest.

The changes are roughly:

- Add common base interface `Pipeline`
- Add translation util
- Use translation util in the executors
- Use PlanExecutors in the streaming environments

## Verifying this change

This is mostly covered by existing tests but we add specific unit tests for the new translation utils.

## Documentation

Only Javadocs, because there are no user-facing changes yet.

### CODE REVIEW

**Comment by aljoscha** [comment]
We should only merge this once FLIP-73 is voted in. I'm opening this now to allow early review/preview.

**Comment by aljoscha** [comment]
@flinkbot run travis

**Comment by TisonKun** [comment]
Give it a quick review here are 2 thoughts

1. where is `FlinkPlan` now?

2. I think although the code conflicts with FLINK-14263 and FLINK-14130 they are not logic conflict. FLINK-14263 and FLINK-14130 are simple code refactors which might be checked in first and this one can rebase on that. Or we can defer the refactor until this one accepted. Either is OK to me.

**Comment by aljoscha** [comment]
Yes, I think FLINK-14263 and FLINK-14130 and can be merged before this.

`FlinkPlan` still exists because some code uses it internally. I could not use this as the "Pipeline" interface because it's internal and there was no existing interface that both `Plan` and `StreamGraph` extend. Or what did you mean by the question?

**Comment by TisonKun** [comment]
Thanks for your ideas on FLINK-14263 and FLINK-14130.

For the issue about `FlinkPlan`, let's revisit the inheritance tree now. We have

```
FlinkPlan
- OptimizedPlan
- StreamingPlan
   - StreamGraph
```

Where `OptimizedPlan` is an instanced class while `StreamingPlan` is an interface, and the "unified interface" `FlinkPlan` is a tag interface.

now with this pull request we have

```
Pipeline
- Plan
- StreamGraph(inherited from StreamingPlan)
```

and two versions of `PipelineTranslator`.

Since `FlinkPlan` is actually a temporary class from batch `Plan` and streaming `StreamGraph` to `JobGraph` and we now have `PipelineTranslator`, I think we can both

1. merge `StreamingPlan` into `StreamGraph` and let `Pipeline` be the unified interface.
2. remove `FlinkPlan`, hide `OptimizedPlan` inside `PlanTranslator` and let it be a more internal concept.

**Comment by aljoscha** [comment]
You are right! I didn't want to go there because it needs more refactoring, mostly around `OptimizerPlanEnvironment`. @kl0u Your refactoring still leaves `OptimizerPlanEnvironment` in place, right?

I'm currently at Flink Forward but will work on this ASAP afterwards. Or you can try based on my PR, if you want, @TisonKun.

**Comment by aljoscha** [comment]
Give me a few minutes, I'm trying it myself right now.

**Comment by TisonKun** [comment]
Thanks for your update @aljoscha ! Go ahead :-)

**Comment by aljoscha** [comment]
I took a bit longer but I think I managed to do it. Also makes the code quite a bit simpler. ☕️

**Comment by aljoscha** [comment]
@TisonKun @kl0u you could have another look now.

File: flink-clients/src/main/java/org/apache/flink/client/FlinkPipelineTranslationUtil.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	/**
+	 * Extracts the execution plan (as JSON) from the given {@link Pipeline}.
+	 */
+	public static String translateToJSONExecutionPlan(Pipeline pipeline) {
+		FlinkPipelineTranslator pipelineTranslator = getPipelineTranslator(pipeline);
+		return pipelineTranslator.translateToJSONExecutionPlan(pipeline);
+	}
+
+	private static FlinkPipelineTranslator getPipelineTranslator(Pipeline pipeline) {
+		PlanTranslator planToJobGraphTransmogrifier = new PlanTranslator();
```
Comment (TisonKun) [inline flink-clients/src/main/java/org/apache/flink/client/FlinkPipelineTranslationUtil.java]:
```suggestion
		PlanTranslator planToJobGraphTranslator = new PlanTranslator();
```

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphTranslator.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	public String translateToJSONExecutionPlan(Pipeline pipeline) {
+		checkArgument(pipeline instanceof StreamGraph,
+				"Given pipeline is not a DataStream StreamGraph.");
+
+		StreamGraph streamGraph = (StreamGraph) pipeline;
+
+		return streamGraph.getStreamingPlanAsJSON();
+	}
+
+	private JobGraph compilePlan(Plan plan, Configuration optimizerConfiguration) {
```
Comment (TisonKun) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphTranslator.java]: Doesn't used anywhere. Also I don't think StreamGraphTranslator should care about batch `Plan`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphTranslator.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+			Configuration optimizerConfiguration,
+			int defaultParallelism) {
+		checkArgument(pipeline instanceof StreamGraph,
+				"Given pipeline is not a DataStream StreamGraph.");
+
+		StreamGraph streamGraph = (StreamGraph) pipeline;
+		return streamGraph.getJobGraph(null);
+	}
+
+	private void setDefaultParallelism(Plan plan, int defaultParallelism) {
```
Comment (TisonKun) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphTranslator.java]: Doesn't used anywhere. Also I don't think StreamGraphTranslator should care about batch Plan

File: flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java [hunk 4]
```diff
@@ -201,18 +157,15 @@ public JobSubmissionResult run(
 		int parallelism,
 		SavepointRestoreSettings savepointSettings) throws CompilerException, ProgramInvocationException {
 
-		OptimizedPlan optPlan = getOptimizedPlan(compiler, plan, parallelism);
```
Comment (TisonKun) [inline flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: With this change `ClusterClient` doesn't have any usage of `compiler`. Also we want to hide `compiler` inside translator. So we can remove `Cluster.compiler` field.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphTranslator.java (hunk 2 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphTranslator.java]: you're right, I thought I found all of these unused code parts 😓

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphTranslator.java (hunk 3 above)
Comment (aljoscha) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphTranslator.java]: Yes, same as above. 👌

File: flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java (hunk 4 above)
Comment (aljoscha) [reply flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java]: you're right, removing it.

**Comment by aljoscha** [comment]
@TisonKun Thanks for the thorough review! ☕️

**Comment by TisonKun** [comment]
Thanks for your update @aljoscha ☕️ nice work!

I think FLINK-14263 has been addressed as a side effect of this pull request. And for a smooth collaboration I'd like to wait for this one get merge and then mark FLINK-14263 as resolved, as well as rebase FLINK-14130(Remove ClusterClient.run() methods) on this patch.

**Comment by aljoscha** [comment]
Thanks! And sorry for "stealing" FLINK-14263" aways from you. 😅

**Comment by TisonKun** [comment]
Never mind 😆 issues can be resolved in many ways

**Comment by TisonKun** [comment]
Hi @aljoscha, is there other blocker of this pull request from merged? I'm going to start subtasks of FLIP-74 port `ClusterClient` to an interface with async methods so that we can implement `JobClient` directly in async version instead of refactor to that after introduce it. As mentioned above, it is somehow conflict on codes with this pull request so I'm waiting on this pull request.

Given that we are the main reviewers of these efforts I don't think I can progress FLIP-74 by myself also :-) If you are lack of time, please let me know the schedule when I can go back to these threads so that I am able to work on other threads without looking this one from time to time.

**Comment by aljoscha** [comment]
You're right, I'm merging this now to unblock future work!

**Comment by TisonKun** [comment]
Thanks for your update Aljoscha. I am proceeding now.

**Pull request closed without merging**